core/trino-main/src/main/java/io/trino/operator/hash/IsolatedHashTableFactory.java

```diff
@@ -0,0 +1,250 @@
+import io.trino.operator.hash.HashTableDataGroupByHash.AddPageWork;
+import io.trino.operator.hash.HashTableDataGroupByHash.GetDictionaryGroupIdsWork;
+import io.trino.operator.hash.HashTableDataGroupByHash.GetGroupIdsWork;
+import io.trino.operator.hash.fixedwidth.FixedWidthEntryStructure;
+import io.trino.operator.hash.fixedwidth.GenericFixedWidthHashTableRowHandler;
+import io.trino.operator.hash.fixedwidth.GenericFixedWidthHashTableValuesAppender;
+import io.trino.operator.hash.fixedwidth.gen.FixedWidthHashTableRowHandler1Channels;
+import io.trino.operator.hash.fixedwidth.gen.FixedWidthHashTableValuesAppender1Channels;
+import io.trino.spi.block.Block;
+import io.trino.sql.gen.IsolatedClass;
+
+import javax.inject.Inject;
+
+import java.lang.reflect.Constructor;
+import java.util.List;
+import java.util.Objects;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class IsolatedHashTableFactory
+{
+    public static final int MAX_SUPPORTED_CHANNELS = 20;
+    private static final Class<? extends HashTableRowHandler>[] DEDICATED_ROW_HANDLERS;
+    private static final Class<? extends HashTableValuesAppender>[] DEDICATED_VALUES_APPENDERS;
+
+    private final LoadingCache<HashTableCacheKey, GroupByHashTable> hashTableCache;
+    private final LoadingCache<FixedWidthEntryStructure, HashTableValuesAppender> appenderCache;
+    private final LoadingCache<StructureAndBlockCacheKey, WorkFactory<Void>> addPageWorkFactoryCache;
+    private final LoadingCache<StructureAndBlockCacheKey, WorkFactory<Void>> addDictionaryPageWorkFactoryCache;
+    private final LoadingCache<StructureAndBlockCacheKey, WorkFactory<GroupByIdBlock>> getGroupIdsWorkFactoryCache;
+    private final LoadingCache<StructureAndBlockCacheKey, WorkFactory<GroupByIdBlock>> getDictionaryGroupIdsWorkFactoryCache;
+
+    static {
+        // setup dedicated per number of channels row handler and values appender classes
+        DEDICATED_ROW_HANDLERS = new Class[MAX_SUPPORTED_CHANNELS];
+        DEDICATED_VALUES_APPENDERS = new Class[MAX_SUPPORTED_CHANNELS];
+        for (int i = 0; i < DEDICATED_ROW_HANDLERS.length; i++) {
+            try {
+                DEDICATED_ROW_HANDLERS[i] = (Class<? extends HashTableRowHandler>) Class.forName(FixedWidthHashTableRowHandler1Channels.class.getPackageName() + "." +
+                        format("FixedWidthHashTableRowHandler%dChannels", i + 1));
+                DEDICATED_VALUES_APPENDERS[i] = (Class<? extends HashTableValuesAppender>) Class.forName(FixedWidthHashTableValuesAppender1Channels.class.getPackageName() + "." +
+                        format("FixedWidthHashTableValuesAppender%dChannels", i + 1));
+            }
+            catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Inject
+    public IsolatedHashTableFactory()
+    {
+        this.hashTableCache = CacheBuilder.newBuilder()
```

Comment:
Do these caches actually help?

Reply:
without it, we would start with interpreted code for every aggregation, every split, even every page. this would both slow the query down and increase pressure on the jit. I did not benchmark it though